Sources/System/FileControl.swift

```diff
@@ -0,0 +1,216 @@
+    public var asPID: CInt? { rawValue >= 0 ? rawValue : nil }
+
+    /// TODO: PGID type
+    @_alwaysEmitIntoClient
+    public var asPositiveGroupID: CInt? {
+      rawValue >= 0 ? nil : -rawValue
+    }
+
+    /// TODO: PID type
+    @_alwaysEmitIntoClient
+    public init(pid id: CInt) {
+      precondition(id >= 0)
+      self.init(rawValue: id)
+    }
+
+    /// TODO: PGID type
+    @_alwaysEmitIntoClient
+    public init(positiveGroupID id: CInt) {
+      precondition(id >= 0)
+      self.init(rawValue: -id)
+    }
+  }
+
+  /// Get the process ID or process group currently receiv-
+  /// ing SIGIO and SIGURG signals.
+  ///
+  /// The corresponding C function is `fcntl` with the `F_GETOWN` command.
+  @_alwaysEmitIntoClient
+  public func getOwner() throws -> PIDOrPGID {
+    try PIDOrPGID(rawValue: fcntl(.getOwner))
+  }
+
+  /// Set the process or process group to receive SIGIO and
+  /// SIGURG signals.
+  ///
+  /// The corresponding C function is `fcntl` with the `F_SETOWN` command.
+  @_alwaysEmitIntoClient
+  public func setOwner(_ id: PIDOrPGID) throws {
+    _ = try fcntl(.setOwner, id.rawValue)
+  }
+}
+
+extension FileDescriptor {
+  /// Duplicate this file descriptor and return the newly created copy.
+  ///
+  /// - Parameters:
+  ///   - `minRawValue`: A lower bound on the new file descriptor's raw value.
+  ///   - `closeOnExec`: Whether the new descriptor's `closeOnExec` flag is set.
+  /// - Returns: The lowest numbered available descriptor whose raw value is
+  ///     greater than or equal to `minRawValue`.
+  ///
+  /// File descriptors are merely references to some underlying system resource.
+  /// The system does not distinguish between the original and the new file
+  /// descriptor in any way. For example, read, write and seek operations on
+  /// one of them also affect the logical file position in the other, and
+  /// append mode, non-blocking I/O and asynchronous I/O options are shared
+  /// between the references. If a separate pointer into the file is desired,
+  /// a different object reference to the file must be obtained by issuing an
+  /// additional call to `open`.
+  ///
+  /// However, each file descriptor maintains its own close-on-exec flag.
+  ///
+  /// The corresponding C functions are `fcntl` with `F_DUPFD` and
+  /// `F_DUPFD_CLOEXEC`.
+  @_alwaysEmitIntoClient
+  public func duplicate(
+    minRawValue: CInt, closeOnExec: Bool
+  ) throws -> FileDescriptor {
+    let cmd: Command = closeOnExec ? .duplicateCloseOnExec : .duplicate
+    return try FileDescriptor(rawValue: fcntl(cmd, minRawValue))
+  }
+
+  // TODO: What is the firmlink thing about?
+
+  /// Get the path of the file descriptor
+  ///
+  /// - Parameters:
+  ///   - `noFirmLink`: Get the non firmlinked path of the file descriptor.
+  ///
+  /// The corresponding C functions are `fcntl` with `F_GETPATH` and
+  /// `F_GETPATH_NOFIRMLINK`.
+  public func getPath(noFirmLink: Bool = false) throws -> FilePath {
```

Comment:
this will need `#if os(Darwin...`. On Linux you do this by running `readlink(/proc/self/FD_NUMBER)`